python/nx-cugraph/nx_cugraph/algorithms/core.py

```diff
@@ -0,0 +1,91 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import numpy as np
+import pylibcugraph as plc
+
+import nx_cugraph as nxcg
+from nx_cugraph.utils import networkx_algorithm, not_implemented_for
+
+__all__ = ["k_truss"]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@networkx_algorithm
+def k_truss(G, k):
+    if is_nx := isinstance(G, nx.Graph):
+        G = nxcg.from_networkx(G, preserve_all_attrs=True)
+    if nxcg.number_of_selfloops(G) > 0:
+        raise nx.NetworkXError(
+            "Input graph has self loops which is not permitted; "
+            "Consider using G.remove_edges_from(nx.selfloop_edges(G))."
+        )
+    # TODO: create renumbering helper function(s)
+    if k < 3:
+        # Drop nodes with zero degree
+        degrees = G._degrees_array()
+        # Renumber step 0: node indices
+        node_indices = degrees.nonzero()[0]
+        if degrees.size == node_indices.size:
+            # No change
+            return G if is_nx else G.copy()
+        src_indices = G.src_indices
+        dst_indices = G.dst_indices
+        # Renumber step 1: edge values (no changes needed)
+        edge_values = {key: val.copy() for key, val in G.edge_values.items()}
+        edge_masks = {key: val.copy() for key, val in G.edge_masks.items()}
+    else:
+        # int dtype for edge_indices would be preferred
+        edge_indices = cp.arange(G.src_indices.size, dtype=np.float64)
+        src_indices, dst_indices, edge_indices, _ = plc.k_truss_subgraph(
+            resource_handle=plc.ResourceHandle(),
+            graph=G._get_plc_graph(edge_array=edge_indices),
+            k=k,
+            do_expensive_check=False,
+        )
+        # Renumber step 0: node indices
+        node_indices = cp.unique(cp.concatenate([src_indices, dst_indices]))
+        # Renumber step 1: edge values
+        edge_indices = edge_indices.astype(np.int64)
+        edge_values = {key: val[edge_indices] for key, val in G.edge_values.items()}
+        edge_masks = {key: val[edge_indices] for key, val in G.edge_masks.items()}
+    # Renumber step 2: edge indices
+    mapper = cp.zeros(len(G), src_indices.dtype)
+    mapper[node_indices] = cp.arange(node_indices.size, dtype=mapper.dtype)
+    src_indices = mapper[src_indices]
+    dst_indices = mapper[dst_indices]
+    # Renumber step 3: node values
+    node_values = {key: val[node_indices] for key, val in G.node_values.items()}
+    node_masks = {key: val[node_indices] for key, val in G.node_masks.items()}
+    # Renumber step 4: key_to_id
+    if (id_to_key := G.id_to_key) is not None:
+        key_to_id = {
+            id_to_key[old_index]: new_index
+            for new_index, old_index in enumerate(node_indices.tolist())
+        }
+    else:
+        key_to_id = None
+    new_graph = G.__class__.from_coo(
```

Comment:
I'm curious why `from_coo` needs to be called this way.

Reply:
I didn't give this much thought or have a strong preference. `from_coo` is a classmethod, so there are three reasonable way to call it:
```python
G.from_coo(...)
G.__class__.from_coo(...)
type(G).from_coo(...)
```
Do you have a preference? Sometimes I think it's weird to read constructors being used off of an instance of a class, hence why I went with `G.__class__`, which is clear it's a class method and is also a little more flexible than `type(G)`. Perhaps this is a "me" thing and I should just use `G.from_coo`. Thoughts?

Reply:
I see, I've always used your option 1 for classmethods (`G.from_coo(...)`), but the pattern you're using is fine by me if you don't think others will get tripped up by it.  I thought it was called that way to somehow get a specific behavior I wasn't familiar with, but your explanation is reasonable to me.

Reply:
I added a code comment about this too :)